lib/Runtime/Library/InJavascript/Intl.js

```diff
@@ -637,113 +654,169 @@
             }
+
+            __fromNLS.set(locale, ret);
+
+            return ret;
         };
 
-        const collationMapForLocale = reverseLocalesAcceptingCollationValues[locale];
-        if (collationMapForLocale === undefined) {
-            // Assume the system wouldn't give us back a bad collation value
-            __mappedDefaultLocale = createLocaleCollationString(locale, collation);
-            return __mappedDefaultLocale;
-        }
+        let __toNLS = new IntlCache();
+        const toNLS = function (locale) {
+            if (!locale) {
+                return undefined;
+            }
 
-        const mappedCollation = collationMapForLocale[collation];
-        if (mappedCollation !== undefined) {
-            __mappedDefaultLocale = createLocaleCollationString(locale, mappedCollation);
-        } else {
-            __mappedDefaultLocale = createLocaleCollationString(locale, collation);
+            if (__toNLS.get(locale)) {
+                return __toNLS.get(locale);
+            }
+
+            const extensionParts = platform.builtInRegexMatch(locale, LANG_TAG_RE);
+            const collationParts = platform.builtInRegexMatch(extensionParts[4] || "", /(?:-co-([^-]+))/);
+            const strippedLocale = platform.builtInRegexMatch(locale, LANG_TAG_BASE_RE)[0];
+            if (!collationParts) {
+                // the given locale had extensions but no -co- extension
+                // NLS doesn't know what to do with anything more than strippedLocale + collation
+                __toNLS.set(locale, strippedLocale);
+                return strippedLocale;
+            }
+
+            const collation = collationParts[1];
+            let ret;
+
+            const collationMapForLocale = localesAcceptingCollationValues[strippedLocale];
+            if (collationMapForLocale !== undefined) {
+                const mappedCollation = collationMapForLocale[collation];
+                if (mappedCollation) {
+                    ret = `${strippedLocale}_${mappedCollation}`;
+                } else {
+                    // if theres no mappedCollation, then NLS doesn't support this collation value (or its the default)
+                    ret = strippedLocale;
+                }
+            } else {
+                // if theres no collationMapForLocale, then NLS doesn't support any collation values for this locale
+                ret = strippedLocale;
+            }
+
+            __toNLS.set(locale, ret);
+
+            return ret;
         }
 
-        return __mappedDefaultLocale;
-    };
+        return [toNLS, fromNLS];
+    })();
 
     // Intl.Collator, String.prototype.localeCompare
     var Collator = (function () {
+        // Keep these "enums" in sync with lib/Runtime/PlatformAgnostic/Intl.h
+        const CollatorSensitivity = bare({
+            BASE: 0,
+            ACCENT: 1,
+            CASE: 2,
+            VARIANT: 3,
+            DEFAULT: 3
```

Comment:
Did we decide to switch to PascalCase on these?

Reply:
We did on the C++ enums. I kept these as all upper case since its easier to normalize arbitrary strings to uppercase than pascal case